File: Back Tracking/组合/90.子集-ii.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;
/*
 * @lc app=leetcode.cn id=90 lang=cpp
 *
 * [90] 子集 II
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> result; // 存放结果
    vector<int> path;
    void backtracking(vector<int> &nums, int startIndex, vector<bool> used)
    {
        result.push_back(path);

        for (int i = startIndex; i < nums.size(); i++)
        {
            if (i > 0 && nums[i - 1] == nums[i] && used[i - 1] == false)
            {
                continue;
            }

            path.push_back(nums[i]);
            used[i] = true;
            // 这里是i+1
            backtracking(nums, i + 1, used);
            used[i] = false;
            path.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int> &nums)
    {

        vector<bool> used(nums.size(), false);
        // 首先把给nums排序，让其相同的元素都挨在一起。
        sort(nums.begin(), nums.end());
        backtracking(nums, 0, used);
        return result;
    }
};
// @lc code=end

File: Back Tracking/组合/90.子集-ii.ts
/*
 * @lc app=leetcode.cn id=90 lang=typescript
 *
 * [90] 子集 II
 */

// @lc code=start
function subsetsWithDup(nums: number[]): number[][] {
  const result: number[][] = []
  const used = new Array(nums.length).fill(false)

  if (nums.length === 0) {
    return result
  }

  // 升序数组，方便做同批次遍历重复元素的判断
  nums = nums.sort((a, b) => a - b)
  const backTracking = (arr: number[], start: number) => {
    result.push([...arr])

    for (let i = start; i < nums.length; i++) {
      // 当前元素与前一个元素相等，同时前一个元素未使用，那么要跳过
      // 例如 1 2 2
      // 同一轮回溯中，第1个 2 和第2个 2在同一个集合中收集
      // 如果非同一轮回溯，1和第2个2组合的情况，实际上1与第1个2组合时候已经包括了，所以要过滤掉
      // used[i-1] = false 时候表示，第1个 2 和第2个 不在同一轮回溯中需要过滤掉
      if (i > 0 && nums[i] === nums[i - 1] && used[i - 1] === false) {
        continue
      }
      // 标记在同一次回溯中使用
      used[i] = true
      arr.push(nums[i])
      backTracking(arr, i + 1)
      // 回溯重置标记为未使用，因为
      used[i] = false
      arr.pop()
    }
  }

  backTracking([], 0)
  return result
};
// @lc code=end

